### DISCUSSION TITLE
Stack 대신 Deque를 사용하는 이유는 무엇일까?

### DISCUSSION THREAD

**Discussion by alstn113** [2024-woowacourse-study/level-interview#32] [Java]
.

**Comment by nayonsoso** [comment]
Stack 대신 Deque를 사용해야 한다는 것은 자바 공식 문서에도 나와있을 만큼 중요한 부분인데요,
그 이유는 바로 Stack 이 Vector를 상속하고 있기 때문입니다.

Stack은 Vector를 상속하기 때문에, 인덱스를 통한 접근이 가능합니다.
이는 LIFO 자료구조에서 기대하지 못한 기능입니다.
따라서 개발자가 실수를 할 여지가 있습니다.

또한, Vector는 동기화된 메서드들로 이루어져있습니다.
때문에 멀티 쓰레드 환경에서 안전하지만,
단일 쓰레드 환경에서는 오버헤드가 발생해 성능이 좋지 않다는 단점이 있습니다.
따라서 단일 쓰레드 환경에서는 Deque를 사용하는 것이 훨씬 성능이 좋습니다.
그럼 멀티 쓰레드 환경에서는 Vector를 사용해야 하는가? 라고 생각하실 수 있는데요,
멀티 쓰레드 환경에 안전한 Deque 구현체를 사용하면 그 문제를 해결할 수 있습니다.

정리하자면, LIFO 구조를 충실히 수행할 수 있으며, 
성능상에도 이점이 있기 때문에 Stack보다 Deque를 사용하는 것이 좋습니다.

**Comment by jimi567** [comment]
Stack이라는 자료구조는 후입선출(Last In Frist Out) 특징을 가지고 있다.
이는 개발자들이 stack이라는 자료구조에 기대하는 특성이기도 하다.
하지만, java의 stack의 경우 내부적으로 vector 를 상속 받고 있어,  인덱스를 통한 접근, 삽입, 제거 등이 실질적으로 가능하게 돼,
Stack의 특징을 유지하지 못하게 된다.

또한, Vector의 메소드 들은 synchronized로 구현되어 있어, 멀티 스레드 환경에서는 동기화의 이점이 있으나,
단일 스레드 환경에서는 불필요한 동기화 작업으로 인해, 성능 측면에서 좋지 않다.

반면에, Deque 인터페이스는 후입선출(LIFO)의 특성을 완전히 유지하면서도 동기화 작업을 가지는 구현체와 그렇지 않은 구현체를 선택할 수 있다. 이는 개발자가 필요에 따라 동기화 작업의 오버헤드를 회피하고 성능을 최적화할 수 있도록 해준다.
